tests/frontend/test_engine.py

```diff
@@ -166,209 +175,299 @@ def inspect():
+        )
 
-        mock__get_blackbird.assert_called_once_with(program=program, param=some_params["param"],
-                args={}, compile_options={})
+        assert connection.get_job_status(id_) == status
 
-        assert result == mock_job.result.result.value
+    def test_get_job_status_error(self, connection, monkeypatch):
+        """Tests a failed job status request."""
+        monkeypatch.setattr(requests, "get", mock_return(mock_response(404, {})))
 
-    def test__run_program_fails(self, starship_engine, monkeypatch):
-        """
-        Tests that an Exception is raised when a job has failed.
-        """
-        mock_to_blackbird = MagicMock()
-        mock__get_blackbird = MagicMock()
-        program = MagicMock()
-        mock_job = MagicMock()
-        mock_job.is_complete = False
-        mock_job.is_failed = True
+        with pytest.raises(RequestFailedError):
+            connection.get_job_status("123")
 
-        monkeypatch.setattr("strawberryfields.engine.to_blackbird", mock_to_blackbird)
-        monkeypatch.setattr(starship_engine, "_get_blackbird", mock__get_blackbird)
-        monkeypatch.setattr(starship_engine, "_queue_job", lambda job_content: mock_job)
+    def test_get_job_result(self, connection, monkeypatch):
+        """Tests a successful job result request."""
+        result_samples = np.array([[1, 2], [3, 4]], dtype=np.int8)
 
-        some_params = {"param": MagicMock()}
+        with io.BytesIO() as buf:
+            np.save(buf, result_samples)
+            buf.seek(0)
+            monkeypatch.setattr(
+                requests,
+                "get",
+                mock_return(mock_response(200, binary_body=buf.getvalue())),
+            )
 
-        with pytest.raises(JobExecutionError) as e:
-            starship_engine._run_program(program, **some_params)
-        assert e.value.args[0] == str(mock_job.manager.http_response_data['meta'])
+        result = connection.get_job_result("123")
 
-    def test__run(self, starship_engine, monkeypatch):
-        """
-        Tests StarshipEngine._run, with the assumption that the backend is a hardware backend
-        that supports running only a single program. This test ensures that a program is compiled
-        for the hardware backend, is locked, is added to self.run_progs, that it is run and that
-        a Result object is returned populated with the result samples.
-        """
+        assert np.array_equal(result.samples.T, result_samples)
+
+    def test_get_job_result_error(self, connection, monkeypatch):
+        """Tests a failed job result request."""
+        monkeypatch.setattr(requests, "get", mock_return(mock_response(404, {})))
+
+        with pytest.raises(RequestFailedError):
+            connection.get_job_result("123")
+
+    def test_cancel_job(self, connection, monkeypatch):
+        """Tests a successful job cancellation request."""
+        monkeypatch.setattr(requests, "patch", mock_return(mock_response(204, {})))
```

Comment:
One thing to note here, that if I understand it correctly any PATCH request is mocked out this way, correct?

One thing that could be checked here, is that `requests.patch` receives the correct arguments. This would include creating a function similar to `mock_return`, but one, that also tracks the arguments passed, such as:
```python

arguments = []
def mock_return(return_value):
    """A helper function for defining a mock function that returns the given value for
    any arguments.
    """
    def function(*args, **kwargs):
        arguments.append(args)
        return return_value
    return function
```
or something comparable.

```python
In [2]: arguments = [] 
   ...: def mock_return(return_value): 
   ...:     """A helper function for defining a mock function that returns the g
   ...: iven value for 
   ...:     any arguments. 
   ...:     """ 
   ...:     def function(*args, **kwargs): 
   ...:         arguments.append(args) 
   ...:         return return_value 
   ...:     return function 
   ...:                                                                         

In [3]: mock_return(3)                                                          
Out[3]: <function __main__.mock_return.<locals>.function(*args, **kwargs)>

In [4]: a = mock_return(3)                                                      

In [5]: a(4, 5, 6)                                                              
Out[5]: 3

In [6]: arguments                                                               
Out[6]: [(4, 5, 6)]
```

Reply:
Yes, any PATCH request with any combination of arguments will return the same response.

That's a good point about validating the arguments. I'll look at adding that into the relevant tests :+1:

Reply:
I've added an argument check to the test - I'm especially grateful that you suggested this because it actually caught a bug!